Wrappers/Python/cil/optimisation/utilities/StepSizeMethods.py

```diff
@@ -138,3 +138,97 @@ def get_step_size(self, algorithm):
+
+    - https://en.wikipedia.org/wiki/Barzilai-Borwein_method
+    """
+
+    def __init__(self, initial, mode='short', stabilisation_param="auto"):
+        '''Initialises the step size rule 
+        '''
+ 
+        self.mode=mode
+        self.store_grad=None 
+        self.store_x=None
+        self.initial=initial
+        if stabilisation_param == 'auto':
+            self.adaptive = True
+            stabilisation_param = numpy.inf
+        elif stabilisation_param == "off":
+            self.adaptive = False 
+            stabilisation_param = numpy.inf
+        elif ( isinstance(stabilisation_param, Number) and stabilisation_param >=0):
+            self.adaptive = False 
+        else:
+            raise TypeError(" The stabilisation_param should be 'auto', a positive number or 'off'")
+        self.stabilisation_param=stabilisation_param
+        
+    
+
+    def get_step_size(self, algorithm):
+        """
+        Applies the B-B rule to calculate the step size (`step_size`)
+
+        Returns
+        --------
+        the calculated step size:float
+
+        """
+        #For the first iteration we use an initial step size because the BB step size requires a previous iterate. 
+        if self.store_x is None:
+            self.store_x=algorithm.x.copy() # We store the last iterate in order to calculate the BB step size 
+            self.store_grad=algorithm.gradient_update.copy()# We store the last gradient in order to calculate the BB step size 
+            return self.initial
+        
+        #If the gradient is zero, gradient based algorithms will not update and te step size calculation will divide by zero so we stop iterations. 
+        if algorithm.gradient_update.norm()<1e-8:
+            raise StopIteration
+
+        if self.mode=='long' or (self.mode =='alternate' and algorithm.iteration%2 ==0):
+            ret = (( algorithm.x-self.store_x).dot(algorithm.x-self.store_x))/ (( algorithm.x-self.store_x).dot(algorithm.gradient_update-self.store_grad))
+        elif self.mode=='short' or (self.mode =='alternate' and algorithm.iteration%2 ==1):
+            ret = (( algorithm.x-self.store_x).dot(algorithm.gradient_update-self.store_grad))/ (( algorithm.gradient_update-self.store_grad).dot(algorithm.gradient_update-self.store_grad))
+        else:
+            raise ValueError('Mode should be chosen from "long", "short" or "alternate". ')
+        
+        #This computes the default stabilisation parameter, using the first three iterations
+        if (algorithm.iteration <=3 and self.adaptive):
+            self.stabilisation_param = min(self.stabilisation_param,(algorithm.x-self.store_x).norm() )
+        
+        # Computes the step size as the minimum of the ret, above, and :math:`\Delta/\|g_k\|` ignoring any NaN values. 
+        ret = numpy.nanmin( numpy.array([ret, self.stabilisation_param/algorithm.gradient_update.norm()]))
```

Comment:
Is `gradient_update.norm`cached? Otherwise store it when you calculate it in `if algorithm.gradient_update.norm()<1e-8:`